ssm_jax/bp/gauss_bp.py

```diff
@@ -0,0 +1,212 @@
+        A22 (dim2, dim2)
+       where `dim1 + dim2 == D` and `dim2 = D - idx`.
+    """
+    split_array = jnp.array([idx])
+    vblocks = jnp.vsplit(A, split_array)
+    # [leaf for tree in forest for leaf in tree]
+    blocks = [block for vblock in vblocks for block in jnp.hsplit(vblock, split_array)]
+    # Can also do:
+    #   blocks = tree_map(lambda arr: jnp.hsplit(arr,split_array),vblocks)
+    # followed by a tree_flatten
+    return blocks
+
+
+def block_join(A11, A12, A21, A22):
+    return jnp.block([[A11, A12], [A21, A22]])
+
+
+def block_rev(A, idx):
+    blocks = block_split(A, idx)
+    return block_join(*blocks[::-1])
+
+
+def vec_swap(a, idx):
+    return jnp.concatenate((a[idx:], a[:idx]))
+
+
+def info_marginalise(K_blocks, hs):
+    """Calculate the parameters of marginalised MVN.
+
+    For x, y joint distributed as
+        p(x, y) = Nc(x,y| h, K),
+    the marginal distribution of x is given by:
+        p(y) = \int p(x, y) dx = Nc(y | hy_marg, Ky_marg)
+    where,
+        hy_marg = h2 - Kyx Kxx^{-1} h1
+        Ky_marg = Kyy - Kyx Kxx^{-1} Kxy
+
+    Args:
+        K_blocks: blocks of the joint precision matrix, (Kxx, Kxy, Kyy),
+                    Kxx (dim_x, dim_x),
+                    Kxy (dim_x, dim_y),
+                    Kyy (dim_y, dim_y).
+        hs (dim_x + dim_y, 1): joint precision weighted mean, (hx, hy):
+                    h1 (dim_x, 1),
+                    h2 (dim_y, 1).
+    Returns:
+        Ky_marg (dim_y, dim_y): marginal precision matrix.
+        hy_marg (dim_y,1): marginal precision weighted mean.
+    """
+    Kxx, Kxy, Kyy = K_blocks
+    hx, hy = hs
+    G = jnp.linalg.solve(Kxx, Kxy)
+    Ky_marg = Kyy - Kxy.T @ G
+    hy_marg = hy - G.T @ hx
+    return Ky_marg, hy_marg
+
+
+def info_condition(Kxx, Kxy, hx, y):
+    """Calculate the parameters of MVN after conditioning.
+
+    For x,y with joint mvn
+        p(x,y) = Nc(x,y | h, K),
+    where h, K can be partitioned into,
+        h = [hx, hy]
+        K = [[Kxx, Kxy],
+            [[Kyx, Kyy]]
+    the distribution of x condition on a particular value of y is given by,
+        p(x|y) = Nc(x | hx_cond, Kx_cond),
+    where
+        hx_cond = hx - Kxy y
+        Kx_cond = Kxx
+    """
+    return Kxx, hx - Kxy @ y
+
+
+def potential_from_conditional_linear_gaussian(A, u, Lambda):
+    """Express a conditional linear gaussian as a potential in canonical form.
+
+    p(y|z) = N(y | Az + u, Lambda^{-1})
+           \prop exp( -0.5(y z)^T K (y z) + (y z)^T h )
+    where,
+        K = (Lambda; -Lambda A,  -A.T Lambda; A.T Lambda A)
+        h = (Lambda u, -A.T Lambda u)
+
+    Args:
+        A (dim1, dim2)
+        u (dim1,1)
+        Lambda (dim1, dim1)
+    Returns:
+        K (dim1 + dim2, dim1 + dim2)
+        h (dim1 + dim2,1)
+    """
+    Kzy = -A.T @ Lambda
+    Kzz = -Kzy @ A
+    Kyy = Lambda
+    hy = Lambda @ u
+    hz = -A.T @ hy
+    return (Kzz, Kzy, Kyy), (hz, hy)
+
+
+def info_multiply(params1, params2):
```

Comment:
Booth `params1` and `params2` are actually tuples (or can be any PyTree) containing all the parameters - in the current uses it tends to be the nested tuple `((K11, K12, K22), (h1, h2))`.  Because all the parameters (K, h, and g if you want) are just summed when two potentials are multiplied, as long as corresponding parameters occupy the same leaves in the PyTree then this will combine them appropriately.
I had been thinking of this flexibility as a virtue but it is perhaps unclear and it might be better to be more explicit about things, at the very least the Docstring could be more descriptive.

Reply:
Yes, just clarify the docstring